src/main/kotlin/org/malachite/estella/services/ApplicationService.kt

```diff
@@ -24,46 +24,65 @@ class ApplicationService(
         offerId: Int,
         jobSeeker: JobSeeker,
         applicationPayload: ApplicationLoggedInPayload
-    ): Application = insertApplication(offerId, jobSeeker, applicationPayload)
+    ): ApplicationDTO = insertApplication(offerId, jobSeeker, applicationPayload)
 
-    fun insertApplication(offerId: Int, jobSeeker: JobSeeker, applicationPayload: ApplicationPayload): Application {
+    fun insertApplication(offerId: Int, jobSeeker: JobSeeker, applicationPayload: ApplicationPayload): ApplicationDTO {
         val offer = offerService.getOffer(offerId)
         val stage = offer.recruitmentProcess?.stages?.getOrNull(0)
         return stage?.let {
             val files = jobSeekerService.addNewFiles(jobSeeker, applicationPayload.getJobSeekerFiles())
             val application = applicationRepository.save(applicationPayload.toApplication(it, jobSeeker, files))
             mailService.sendApplicationConfirmationMail(offer, application)
-            application
+            addNewApplicationStage(application, it).toApplicationDTO()
         } ?: throw UnsupportedOperationException("First stage not found in application")
     }
 
-    fun insertApplicationWithoutUser(offerId: Int, applicationPayload: ApplicationNoUserPayload): Application =
+    fun insertApplicationWithoutUser(offerId: Int, applicationPayload: ApplicationNoUserPayload): ApplicationDTO =
         jobSeekerService.getOrCreateJobSeeker(applicationPayload.toJobSeeker())
             .let { insertApplication(offerId, it, applicationPayload) }
 
 
-    fun setNextStageOfApplication(applicationId: Int) {
+    fun setNextStageOfApplication(applicationId: Int, recruitmentProcess: RecruitmentProcess) {
         val application = applicationRepository.findById(applicationId).get()
 
         if (application.status != ApplicationStatus.IN_PROGRESS)
             throw UnsupportedOperationException("Cannot change stage of resolved application!")
 
-        val recruitmentProcessStages = recruitmentProcessService
-            .getProcessFromStage(application.stage)
+        val recruitmentProcessStages = recruitmentProcess
             .stages
             .sortedBy { it.id }
-        val index = recruitmentProcessStages.indexOf(application.stage)
-        if (index == recruitmentProcessStages.lastIndex - 1)
-            applicationRepository.save(
-                application.copy(
-                    stage = recruitmentProcessStages[index + 1],
-                    status = ApplicationStatus.ACCEPTED
-                )
-            )
-        else if (index < recruitmentProcessStages.lastIndex)
-            applicationRepository.save(application.copy(stage = recruitmentProcessStages[index + 1]))
+
+        val applicationStages = application.applicationStages.map { it.stage }.sortedBy { it.id }
+
+        if (applicationStages.isEmpty()) {
```

Comment:
Trochę nie rozumiem, co się dzieje w tym kawałku kodu (tu i w dół)

Reply:
1. Z wszystkich applicationStages applikacji bierzemy recruitmentStage i sortujemy po id. 
2. Jeśli nie ma żadnego recruitmentStage (ergo także applicationStage) to tworzym applicationStage z odwołaniem do pierwszego stage z recruitmentProcessu.
3. Jeśli istnieją jakieś applicationStage to sprawdzamy czy istnieje jakiś następny RecruitmentStage którego nie ma w tej aplikacji jest jest to dodajemy, dodatkowo sprawdzamy czy ten stage do którego przeszliśmy nie jest ostani jeśli jest to zmieniamy status aplikacji na accepted.

Reply:
Mogę to przepisać ale przedtem to wyglądało podobnie przynajmniej tak mi się wydaje.